#In this homework assignment, we apply Multiple Linear Regression (MLR) to a real-world dataset that includes missing values and redundant information.

In [ ]:
#@title Mounting your Google Drive
#from google.colab import drive
#drive.mount('/content/drive/', force_remount=True)

Mounted at /content/drive/


In [1]:
#@title Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [2]:
# Import the dataset (name it 'dataset') you have uplouaded into your Google Drive
# Hint: use pandas AND read_csv

dataset = pd.read_csv('Car_DS_original_v3.csv')

In [3]:
# Run this cell to see the first five rows of the dataset
# Note: if you want to see more rows, use DS.head(number of rows)
dataset.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner     mileage   engine   max_power  \
0  Individual       Manual   First Owner   23.4 kmpl  1248 CC      74 bhp   
1  Individual       Manual  Second Owner  21.14 kmpl  1498 CC  103.52 bhp   
2  Individual       Manual   Third Owner   17.7 kmpl  1497 CC      78 bhp   
3  Individual       Manual   First Owner   23.0 kmpl  1396 CC      90 bhp   
4  Individual       Manual   First Owner   16.1 kmpl  1298 CC    88.2 bhp   

             torque  seats  
0       190Nm@ 2000    5.0  
1      250Nm@  2500    5.0  
2        12.7@ 2700    5.0  
3  22.4 kgm @  2750    5.0  
4        11.5@ 4500    5.0

In [4]:
# Run this cell to see dataset information
# Expected output is shown is the text cell below.
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8128 entries, 0 to 8127
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           8128 non-null   object 
 1   year           8128 non-null   int64  
 2   selling_price  8128 non-null   int64  
 3   km_driven      8128 non-null   int64  
 4   fuel           8128 non-null   object 
 5   seller_type    8128 non-null   object 
 6   transmission   8128 non-null   object 
 7   owner          8128 non-null   object 
 8   mileage        7907 non-null   object 
 9   engine         7907 non-null   object 
 10  max_power      7913 non-null   object 
 11  torque         7906 non-null   object 
 12  seats          7907 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 825.6+ KB


In [5]:
# Run this cell to see the first five rows of the cleaned data
dataset.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner     mileage   engine   max_power  \
0  Individual       Manual   First Owner   23.4 kmpl  1248 CC      74 bhp   
1  Individual       Manual  Second Owner  21.14 kmpl  1498 CC  103.52 bhp   
2  Individual       Manual   Third Owner   17.7 kmpl  1497 CC      78 bhp   
3  Individual       Manual   First Owner   23.0 kmpl  1396 CC      90 bhp   
4  Individual       Manual   First Owner   16.1 kmpl  1298 CC    88.2 bhp   

             torque  seats  
0       190Nm@ 2000    5.0  
1      250Nm@  2500    5.0  
2        12.7@ 2700    5.0  
3  22.4 kgm @  2750    5.0  
4        11.5@ 4500    5.0

#In the sequel, we aim to clean the dataset column that includes both letters and numbers, ensuring it contains only numbers.

In [6]:
# Start with 'torque' column. We only need the part after '@' sign in this column
# The following will split each string at '@' and keep only the part after '@'
dataset['torque'] = dataset['torque'].str.split('@').str[1]

# Save the cleaned data, i.e., dataset_cleaned, to a new file
# Hint: use to_csv command
dataset.to_csv('dataset_cleaned.csv', index=False)

In [7]:
# Run this cell to see the first five rows of the cleaned data
dataset.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner     mileage   engine   max_power  \
0  Individual       Manual   First Owner   23.4 kmpl  1248 CC      74 bhp   
1  Individual       Manual  Second Owner  21.14 kmpl  1498 CC  103.52 bhp   
2  Individual       Manual   Third Owner   17.7 kmpl  1497 CC      78 bhp   
3  Individual       Manual   First Owner   23.0 kmpl  1396 CC      90 bhp   
4  Individual       Manual   First Owner   16.1 kmpl  1298 CC    88.2 bhp   

   torque  seats  
0    2000    5.0  
1    2500    5.0  
2    2700    5.0  
3    2750    5.0  
4    4500    5.0

In [8]:
# Remove 'bhp' from the 'max_power' column and replace with space
dataset['max_power'] = dataset['max_power'].str.replace('bhp', '').str.strip()

# Convert the 'max_power' column to a numeric type (float)
dataset['max_power'] = pd.to_numeric(dataset['max_power'], errors='coerce')

# Verify the changes by printing the first five rows of the revised dataset
print(dataset.head())


                           name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner     mileage   engine  max_power  \
0  Individual       Manual   First Owner   23.4 kmpl  1248 CC      74.00   
1  Individual       Manual  Second Owner  21.14 kmpl  1498 CC     103.52   
2  Individual       Manual   Third Owner   17.7 kmpl  1497 CC      78.00   
3  Individual       Manual   First Owner   23.0 kmpl  1396 CC      90.00   
4  Individual       Manual   First Owner   16.1 kmpl  1298 CC      88.20   

   torque  seats  
0    2000    5.0  
1    2500    5.0  
2    2700    5.0  
3    2750    5.0

In [9]:
# Remove 'kmpl' from the 'mileage' column and replace with space
dataset['mileage'] = dataset['mileage'].str.replace('kmpl', '').str.strip()

# Convert the 'mileage' column to a numeric type (float)
dataset['mileage'] = pd.to_numeric(dataset['mileage'], errors='coerce')

# Verify the changes by printing the first five rows of the revised dataset
print(dataset.head())


                           name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner  mileage   engine  max_power  torque  \
0  Individual       Manual   First Owner    23.40  1248 CC      74.00    2000   
1  Individual       Manual  Second Owner    21.14  1498 CC     103.52    2500   
2  Individual       Manual   Third Owner    17.70  1497 CC      78.00    2700   
3  Individual       Manual   First Owner    23.00  1396 CC      90.00    2750   
4  Individual       Manual   First Owner    16.10  1298 CC      88.20    4500   

   seats  
0    5.0  
1    5.0  
2    5.0  
3    5.0  
4    5.

In [ ]:
# Remove 'CC' from the 'engine' column and replace with space
dataset['engine'] = dataset['engine'].str.replace('CC', '').str.strip()

# Convert the 'engine' column to a numeric type (float)
dataset['engine'] = pd.to_numeric(dataset['engine'], errors='coerce')

# Verify the changes by printing the first five rows of the revised dataset
print(dataset.head())


In [10]:
# Check if there is any null values in the 'dataset_cleaned'
# Hint: use isna() command
# Expected output is shown is the text cell below.

print(dataset.isna().sum())

name               0
year               0
selling_price      0
km_driven          0
fuel               0
seller_type        0
transmission       0
owner              0
mileage          221
engine           221
max_power        216
torque           263
seats            221
dtype: int64


In [11]:
# Drop rows with any column having NA/null data
# Hint: use dropna() command
# Name dataset with no NA/null data as 'dataset_cleaned'
dataset_cleaned = dataset.dropna()

# Verify the changes by printing the first five rows of the cleaned dataset
print(dataset_cleaned.head())


                           name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner  mileage   engine  max_power  torque  \
0  Individual       Manual   First Owner    23.40  1248 CC      74.00    2000   
1  Individual       Manual  Second Owner    21.14  1498 CC     103.52    2500   
2  Individual       Manual   Third Owner    17.70  1497 CC      78.00    2700   
3  Individual       Manual   First Owner    23.00  1396 CC      90.00    2750   
4  Individual       Manual   First Owner    16.10  1298 CC      88.20    4500   

   seats  
0    5.0  
1    5.0  
2    5.0  
3    5.0  
4    5.

In [12]:
# Save the cleaned data, i.e., dataset_cleaned, to a new file
dataset_cleaned.to_csv('dataset_cleaned.csv', index=False)


In [13]:
# Select feature columns (feature matrix) by selecting all columns except
# the 'selling_price', 'seller_type', 'name', 'owner'. Name them 'X' as it would be a matrix
# Hint: you can select columns by their names or drop the column that you don't want.

X = dataset_cleaned.drop(columns=['selling_price', 'seller_type', 'name', 'owner'])

print(X.head())


   year  km_driven    fuel transmission  mileage   engine  max_power  torque  \
0  2014     145500  Diesel       Manual    23.40  1248 CC      74.00    2000   
1  2014     120000  Diesel       Manual    21.14  1498 CC     103.52    2500   
2  2006     140000  Petrol       Manual    17.70  1497 CC      78.00    2700   
3  2010     127000  Diesel       Manual    23.00  1396 CC      90.00    2750   
4  2007     120000  Petrol       Manual    16.10  1298 CC      88.20    4500   

   seats  
0    5.0  
1    5.0  
2    5.0  
3    5.0  
4    5.0  


In [14]:
# Select the column that you want to predict, i.e., 'selling_price'
# Name it 'y' as it would be a vector.
y = dataset_cleaned['selling_price']

print(y.head())


0    450000
1    370000
2    158000
3    225000
4    130000
Name: selling_price, dtype: int64


# Encoding categorical data

In [40]:
# You have to code catgorical data. To do so, use the following libraries.
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer


#Preprocess 'engine' column to remove 'CC' and convert to numeric
X['engine'] = X['engine'].str.replace(' CC', '').astype(float)
# First step: select dataset columns that you want to code
categorical_feature = ['transmission', 'fuel'] 
# Second step: defining the encoder
one_hot = OneHotEncoder()

# Third step: Define transformer to apply encoder to the selected dataset columns
transformer = ColumnTransformer(
    transformers=[('cat', one_hot, categorical_feature)], remainder='passthrough')

# Fourth step: Fit the defined transformer to the feature matrix X
transformed_X = transformer.fit_transform(X)


In [41]:
# Run this cell to see the first five rows of transformed_X
pd.DataFrame(transformed_X).head()

0    1    2    3    4    5     6       7      8       9       10      11  \
0  0.0  1.0  0.0  1.0  0.0  0.0  2014  145500   23.4  1248.0    74.0    2000   
1  0.0  1.0  0.0  1.0  0.0  0.0  2014  120000  21.14  1498.0  103.52    2500   
2  0.0  1.0  0.0  0.0  0.0  1.0  2006  140000   17.7  1497.0    78.0    2700   
3  0.0  1.0  0.0  1.0  0.0  0.0  2010  127000   23.0  1396.0    90.0    2750   
4  0.0  1.0  0.0  0.0  0.0  1.0  2007  120000   16.1  1298.0    88.2    4500   

    12  
0  5.0  
1  5.0  
2  5.0  
3  5.0  
4  5.0

# Splitting the dataset into the Training set and Test set

In [42]:
# Split the dataset into the Training and Test sets using the
# following imported library;
# Consider test size to be 0.35 and random state to be 2509
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(transformed_X, y, test_size=0.35, random_state=2509)

pd.DataFrame(X_train).head()

0    1    2    3    4    5     6      7      8       9      10      11  \
0  0.0  1.0  0.0  0.0  0.0  1.0  2018  15000   24.7   796.0   47.3    3500   
1  1.0  0.0  0.0  1.0  0.0  0.0  2014  51000  14.74  2993.0  270.9    2000   
2  0.0  1.0  0.0  0.0  0.0  1.0  2017  40000  23.95   998.0  67.05    3500   
3  1.0  0.0  0.0  0.0  0.0  1.0  2019   1500   18.2  1199.0   88.7    4800   
4  0.0  1.0  0.0  0.0  0.0  1.0  2016  44000  25.17   799.0   53.3   4386    

    12  
0  5.0  
1  5.0  
2  5.0  
3  5.0  
4  5.0

In [43]:
# This cell standardized training and test data

# Initialize the scaler
scaler = StandardScaler()

# Fit and transform the training data
X_train_scaled = scaler.fit_transform(X_train)

# Transform the test data
X_test_scaled = scaler.transform(X_test)

# Training the Multiple Linear Regression model on the Training set

In [44]:
# Train the Multiple Linear Regression model on the Training set

# First step: import linear regression from sklearn
from sklearn.linear_model import LinearRegression

# Second step: define regeressor using the imported linear regression
regressor = LinearRegression()

# Fit the defined regeressor to the training dataset, including X_train_scaled and y_train
regressor.fit(X_train_scaled, y_train)

LinearRegression()

In [45]:
# Evaulate the performance of your model on X_test_scaled and y_test.
# In the context of multilinear regression),
# Hint: use .score() method to return the coefficient of determination, also known as R² (R-squared).
# We expect output > 0.63
r_squared = regressor.score(X_test_scaled, y_test)
print(f"R^2 (Coefficient of Determination): {r_squared}")

R^2 (Coefficient of Determination): 0.6371409668462373


In [46]:
# Apply cross validation (CV)

# Import CV score from sklearn
from sklearn.model_selection import cross_val_score

# Apply K-Fold cross-validation
# Using K=5 and negative mean squared error for scoring
cv_scores = cross_val_score(regressor, X_train_scaled, y_train, cv=5, scoring='neg_mean_squared_error')

# Print the score mean (± standard deviation)
print(f"Mean CV Score: {cv_scores.mean()} (± {cv_scores.std()})")


Mean CV Score: -203893477098.84824 (± 10887305221.16819)


# Predicting the Test set results

In [47]:
# Apply the regressor on X_test_scaled to predict y_pred
y_pred = regressor.predict(X_test_scaled)

In [48]:
# This cell shows the actual label vector y_test and the predicted label vector y_pred
d = {'y_pred': y_pred, 'y_test': y_test}
pd.DataFrame(d)

y_pred   y_test
3541  3.112460e+05   660000
4740  4.108823e+05   500000
3759  9.040131e+05   450000
4919 -1.211936e+05   175000
5301  4.809011e+05   482000
...            ...      ...
6599  8.043051e+05   640000
7786  1.405673e+05   350000
299   4.269944e+05   525000
57    2.264055e+06  1750000
3093  2.702371e+06  5500000

[2753 rows x 2 columns]

In [49]:
# As an example metric, run this cell to determine Mean Squared Error
mse = mean_squared_error(y_test, y_pred)
print("Mean Squared Error:", mse)

Mean Squared Error: 224403746956.12976


## **(2) Ridge Regression**

In [50]:
# Apply the Ridge Regression

# First step: import Ridge from sklearn
from sklearn.linear_model import LinearRegression, Ridge

# Define ridge_model based on the imported Ridge; consider alpha=1
# alpha is the regularization strength
ridge_model = Ridge(alpha=1)

# Fit the ridge_model to the X_train_scaled and y_train
ridge_model.fit(X_train_scaled, y_train)

Ridge(alpha=1)

In [51]:
# Use Ridge Regression for prediction
# Apply ridge_model to X_test_scaled
y_pred_ridge = ridge_model.predict(X_test_scaled)

# Determine mean squared error of the ridge_model
# Between y_test and y_pred_ridge
from sklearn.metrics import mean_squared_error
mse_ridge = mean_squared_error(y_test, y_pred_ridge)

# Print mse_ridge
print(mse_ridge)


224405754176.9037


In [52]:
# Apply K-Fold cross-validation to ridge_model
# Consider K=5, and use negative mean squared error for scoring
ridge_cv_scores = cross_val_score(ridge_model, X_train_scaled, y_train, cv=5, scoring='neg_mean_squared_error')

# Print CV MSE for Ridge Regression, i.e., the score mean (± standard deviation)
print(f"Mean CV Score for Ridge Regression: {ridge_cv_scores.mean()} (± {ridge_cv_scores.std()})")


Mean CV Score for Ridge Regression: -203887847392.03976 (± 10906563605.55444)


## **(3) Lasso Regression**

In [53]:
# Apply the Lasso Regression

# First step: import Lasso from sklearn
from sklearn.linear_model import Lasso

# Define lasso_model based on the imported Lasso; consider alpha=0.25
# alpha is the regularization strength
lasso_model = Lasso(alpha=0.25)

# Fit the lasso_model to the X_train_scaled and y_train
lasso_model.fit(X_train_scaled, y_train)


Lasso(alpha=0.25)

In [54]:
# Use Lasso Regression for prediction
# Apply lasso_model to X_test_scaled
y_pred_lasso = lasso_model.predict(X_test_scaled)

# Determine mean squared error of the lasso_model
# Between y_test and y_pred_lasso
from sklearn.metrics import mean_squared_error
mse_lasso = mean_squared_error(y_test, y_pred_lasso)

# Print mse_lasso
print(mse_lasso)

224403739518.70273


In [55]:
# Apply K-Fold cross-validation to lasso_model
# Consider K=5, and use negative mean squared error for scoring
lasso_cv_scores = cross_val_score(lasso_model, X_train_scaled, y_train, cv=5, scoring='neg_mean_squared_error')

# Print CV MSE for Lasso Regression, i.e., the score mean (± standard deviation)
print(f"Mean CV Score for Lasso Regression: {lasso_cv_scores.mean()} (± {lasso_cv_scores.std()})")


Mean CV Score for Lasso Regression: -203893449070.6928 (± 10887414248.643572)


In [57]:
# Check Coefficients correspond to each feature;
# Larger cofficient shows more impact of that feature on the output

print(f'Coefficients of the original model are: {regressor.coef_}')
print(f'Intercept of the original model is: {regressor.intercept_}')

print(f'\n Coefficients of the Ridge model are: {ridge_model.coef_}')
print(f'Intercept of the Ridge model is: {ridge_model.intercept_}')

print(f'\n Coefficients of Ridge model are: {lasso_model.coef_}')
print(f'Intercept of the Lasso model is: {lasso_model.intercept_}')

Coefficients of the original model are: [  77929.17633664  -77929.17633663    4675.68854254  -41144.17340948
   19172.20674324   37923.14096455  125085.13410428  -82813.82606474
   61530.12773975   79973.22964392  476091.63912567 -122341.52290476
  -39413.3684155 ]
Intercept of the original model is: 638674.8282472423

 Coefficients of the Ridge model are: [  77959.32511204  -77959.32511192    4666.27797958  -41054.34986826
   19155.02723891   37837.03094026  125091.42104472  -82832.80777395
   61466.98452499   80051.26256371  475841.85368317 -122202.1792132
  -39459.18896508]
Intercept of the Ridge model is: 638674.8282472423

 Coefficients of Ridge model are: [ 1.55858408e+05 -9.46046710e-10 -2.13581571e+03 -7.87553899e+04
  1.43452241e+04  3.95056489e+02  1.25084994e+05 -8.28138672e+04
  6.15291497e+04  7.99714086e+04  4.76091559e+05 -1.22340076e+05
 -3.94129076e+04]
Intercept of the Lasso model is: 638674.8282472424


#In this homework, we observed that Multiple Linear Regression (MLR) is not fully capable of learning the model, achieving a prediction score of around 0.64. In future sessions, we will introduce RandomForest and demonstrate how it can significantly improve performance.